CORE-4868 - sandbox changes for force cpi upload

conal-smith opened the pull request:
The main changes to look at first:
- CpkReadService API change to the get function and for implementation change the key of the map to checksum of CPK.
- VirtualNodeContext now has cpkFileChecksums rather than CPKIdentifiers because they uniquely identify files loaded from disk by CpkReadService.
- workers-smoketest build.gradle now builds two calculators and makes assertions about these in VirtualNodeRpcTest, including ability to run the new multiplication calculator, which verifies the flow sandbox cache has been updated and we're able to run the correct flow

See the following illustration:
![image](https://user-images.githubusercontent.com/68279385/175280252-63997dc2-70c9-4ebf-99e9-2ae246854519.png)

sprintengine commented on components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/DatabaseChunkPersistence.kt:
```diff
@@ -229,21 +233,16 @@ class DatabaseChunkPersistence(private val entityManagerFactory: EntityManagerFa
-            // persist metadata
+            val cpkMetadataEntities = createOrUpdateCpkMetadataEntities(cpi, findExistingCpiCpks(em, cpi))
+
+            val cpiMetadataEntity = createCpiMetadataEntity(cpi, cpiFileName, checksum, requestId, groupId, cpkMetadataEntities)
```
Changes to `persistMetadataAndCpks` allows fetching existing CpiCpk entities so when we merge them, their `@Version` is preserved.

Also preserves version for CpkMetadataEntity and CpkFileEntity.

This could be extracted to a separate PR.

sprintengine commented on components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseChunkPersistenceTest.kt:
```diff
@@ -657,10 +657,39 @@ internal class DatabaseChunkPersistenceTest {
+
+        persistence.updateMetadataAndCpks(anotherUpdatedCpi, "test.cpi", newRandomSecureHash(), UUID.randomUUID().toString(), "group-b")
+
+        findAndAssertCpk(cpi.metadata.cpiId, cpk.metadata.cpkId, thirdChecksum.toString(), 2, 4, 2)
```
I found that https://github.com/corda/corda-runtime-os/pull/1483 actually lacked a test that made sure the entityVersion was copied over when CpiCpk was > 1. I found an issue using kotlin copy() function where some fields weren't being copied, including the version field and insertTimestamp.

For now, I'm just creating new objects instead of "copying" data classes. There is a separate PR to see if we can use `em.getReference()`  and I'd prefer to get this jira finished with this quick fix and fix it up later with getReference or possible remove kotlin data classes. https://r3-cev.atlassian.net/browse/CORE-5271

sprintengine commented on libs/packaging/packaging-core/src/main/kotlin/net/corda/libs/packaging/core/CpkIdentifier.kt:
```diff
@@ -41,7 +41,7 @@ data class CpkIdentifier(
                 )
             },
-            0
+            0//todo conal - remove this
```
No longer need entityVersion on CpkIdentifier avro object, so I'll remove it later.

vkolomeyko replied:
Perhaps, we should raise a dedicated Jira for that, as this is likely to require `corda-api` change.

sprintengine replied:
Created https://r3-cev.atlassian.net/browse/CORE-5337

sprintengine commented on libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiCpkEntity.kt:
```diff
@@ -60,11 +60,16 @@ data class CpiCpkEntity(
         metadata: CpkMetadataEntity,
     ): CpiCpkEntity {
-        return this.copy(
+        val copy = CpiCpkEntity(
```
Here is where I saw using this.copy was not copying over entityVersion or insertTimestamp.

sprintengine replied:
I'd quite like to improve these, I'll look into it.

sprintengine commented on libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiMetadataEntity.kt:
```diff
@@ -138,21 +138,7 @@ data class CpiMetadataEntity(
             fileUploadRequestId = fileUploadRequestId,
             fileName = fileName,
             fileChecksum = fileChecksum,
-            cpks = cpks.map { (originalFileName, cpiCpkEntity) ->
```
This also shouldn't have been creating new CpiCpk from the newly updated objects we passed in. But still, the new cpiCpk object didn't have insertTimestamp or entityVersion on it due to above copy issue.

driessamyn replied:
That's because copy only copies things that are in the data class' constructor.
Can we not do a getReference here if we need the managed entity rather than create a new one?

sprintengine commented on libs/virtual-node/sandbox-group-context/src/main/kotlin/net/corda/sandboxgroupcontext/VirtualNodeContext.kt:
```diff
@@ -12,7 +12,7 @@ import net.corda.virtualnode.HoldingIdentity
 data class VirtualNodeContext(
     val holdingIdentity: HoldingIdentity,
-    val cpkIdentifiers: Set<CpkIdentifier>,
+    val cpkFileChecksums: Set<SecureHash>,
```
Here is part of the crux of the change, we identify the CPKs this context has by their cpkFileChecksum as this is a natural unique identifier and we make sure we're targetting the correct cpk when it comes to finding the CPK from CpkReadService.

sprintengine commented on testing/cpbs/calculator-for-cache-invalidation-testing/src/main/kotlin/net/cordapp/testing/calculator/CalculatorFlow.kt:
```diff
@@ -0,0 +1,42 @@
+import net.corda.v5.base.annotations.Suspendable
+import net.corda.v5.base.util.contextLogger
+
+class CalculatorFlow : RPCStartableFlow {
```
Multiplication calculator, produces exactly the same artifact as addition calculator so we can "force" upload it to overwrite the existing calculator.

sprintengine commented on components/virtual-node/cpk-read-service/src/main/kotlin/net/corda/cpk/read/CpkReadService.kt:
```diff
@@ -1,9 +1,9 @@
 
 interface CpkReadService : Lifecycle {
-    fun get(cpkId: CpkIdentifier): Cpk?
+    fun get(cpkFileChecksum: SecureHash): Cpk?
```
Change to this API is important, allows us to correctly identify CPKs, now that multiple can have same cpkId, but they'll always have different fileChecksums

sprintengine commented on components/virtual-node/cpk-read-service-impl/src/main/kotlin/net/corda/cpk/read/impl/CpkReadServiceImpl.kt:
```diff
@@ -80,7 +80,7 @@ class CpkReadServiceImpl (
     internal var cpkChunksKafkaReaderSubscription: AutoCloseable? = null
 
-    private val cpksById = ConcurrentHashMap<CpkIdentifier, Cpk>()
+    private val cpksByChecksum = ConcurrentHashMap<SecureHash, Cpk>()
```
And changing this map to be keyed off the cpkFileChecksum to allow the API change.

sprintengine commented on libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkMetadataEntity.kt:
```diff
@@ -36,12 +36,15 @@ data class CpkMetadataEntity(
             serializedMetadata = serializedMetadata,
             isDeleted = isDeleted
         )
+        copy.entityVersion = entityVersion
```
I think this entity should have var fields i.e. cpkFileChecksum, formatVersion, serializedMetadata, and id should be a val

sprintengine commented on libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkFileEntity.kt:
```diff
@@ -45,11 +45,14 @@ data class CpkFileEntity(
         data: ByteArray,
```
fileChecksum, data, (insertTimestamp?) should be var fields, id should be a val

vkolomeyko commented on applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/virtualnode/VirtualNodeRpcTest.kt:
```diff
@@ -222,76 +223,138 @@ class VirtualNodeRpcTest {
+            // Cannot use `assertWithRetry` as there is a strict type `Instant`
+            // Allow ample time for CPI upload to be propagated through the system
+            eventually(Duration.ofSeconds(20)) {
+                assertThat(getCpkTimestamp()).isAfter(initialCpkTimeStamp)
```
I wonder if we can assert `entityVersion` being incremented here as well?

sprintengine replied:
We'd have to add entityVersion onto CpkMetadata (cpi-upload-endpoints) as well as CpkMetadata (packaging-core) and the corresponding avro object CpkMetadata. However, maybe this will be required anyway when we move on to vnode lifecycle update, entityVersion may be required on the message bus for optimistic locking. I think for this PR, its enough to be without it, and maybe add in a separate PR or separate jira. Wdyt?

driessamyn replied:
I think it's probably only the CPI entity version that is relevant as that is what you push. In terms of the API, the CPK entity version isn't that relevant, but then having it on the API may be useful from a troubleshooting point of view.
My feeling is, if we add it, probably under a separate JIRA/PR is fine.

driessamyn commented on applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/virtualnode/VirtualNodeRpcTest.kt:
```diff
@@ -222,76 +223,138 @@ class VirtualNodeRpcTest {
+
+        val requestBody = """{ "requestBody":  "{ \"a\":$a, \"b\":$b }" }"""
+
+        // BUG:  Due to a bug in the flow worker, we *still* have to wait for a while for the CPI to be assembled
+        // otherwise the flow-worker just stops responding if a flow is requested before the CPI.
+        Thread.sleep(FLOW_WAIT_DURATION.toMillis())
```
Bit unrelated to this PR but it might be worth checking with @owenstanford if this is still needed as I think it was being addressed and it would save you approx 60 seconds each time you run the test ;)

sprintengine replied:
It seems to work locally without this, so I'll commit it and see

conal-smith commented:
Just fixing up a few issues post-rebase with recent changes to flow db integration

driessamyn commented on components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/DatabaseChunkPersistence.kt:
```diff
@@ -234,12 +234,28 @@ class DatabaseChunkPersistence(private val entityManagerFactory: EntityManagerFa
 
-            val cpiMetadataEntity = createCpiMetadataEntity(cpi, cpiFileName, checksum, requestId, groupId, cpkMetadataEntities)
-            
+            val cpiCpkEntities = cpi.cpks.mapTo(HashSet()) { cpk ->
```
A bit outside the scope of this change, but could we take the opportunity to rename these methods? `persist` and `update` aren't distinctive enough. The original intention was `persist` could be generic (create or update), but that's not how it's panned out, so let's make it explicit.

sprintengine replied:
Yeah I have a separate task https://r3-cev.atlassian.net/browse/CORE-5271 and I'll add this comment as a todo as well.

barrylapthorn replied:
FWIW I renamed this class to `CpiPersistence` before I reverted it yesterday.

driessamyn commented on components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/DatabaseChunkPersistence.kt:
```diff
@@ -234,12 +234,28 @@ class DatabaseChunkPersistence(private val entityManagerFactory: EntityManagerFa
+                    CpkMetadataEntity(
+                        cpk.metadata.cpkId.toCpkKey(),
+                        cpk.metadata.fileChecksum.toString(),
+                        "1",
```
format version needs to be taken from the metadata: `cpk.metadata.manifest.cpkFormatVersion`

(this was relatively recently added) - FYI @josephzunigadaly

sprintengine replied:
Nice one thanks!

driessamyn commented on components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/DatabaseChunkPersistence.kt:
```diff
@@ -352,57 +369,38 @@ class DatabaseChunkPersistence(private val entityManagerFactory: EntityManagerFa
+                entityToUpdate.isDeleted = false
+                entityToUpdate.metadata.cpkFileChecksum = cpk.metadata.fileChecksum.toString()
+                entityToUpdate.metadata.serializedMetadata = cpk.metadata.toJsonAvro()
+                entityToUpdate.metadata.formatVersion = "1"
```
`cpk.metadata.manifest.cpkFormatVersion`

driessamyn commented on components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/DatabaseChunkPersistence.kt:
```diff
@@ -352,57 +369,38 @@ class DatabaseChunkPersistence(private val entityManagerFactory: EntityManagerFa
+                    CpkMetadataEntity(
+                        cpkKey,
+                        cpk.metadata.fileChecksum.toString(),
+                        "1",
```
`cpk.metadata.manifest.cpkFormatVersion`

driessamyn commented on components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/DatabaseChunkPersistence.kt:
```diff
@@ -418,15 +416,17 @@ class DatabaseChunkPersistence(private val entityManagerFactory: EntityManagerFa
-                ?: CpkFileEntity(
+            if (existingCpkFileEntity != null) {
+                existingCpkFileEntity.fileChecksum = cpk.metadata.fileChecksum.toString()
+                existingCpkFileEntity.data = Files.readAllBytes(cpk.path!!)
```
This doesn't make sense I think. This assumes that the "new" CPKs are different to the existing, but that isn't always the case.  I think you need to do a checksum equality test first to establish this.
Otherwise it could be a bit wasteful in case of large CPK that weren't updated (think about somebody referencing something like Tokens SDK).

sprintengine replied:
Yeah good point. I'll change to add a fileChecksum equality check. But this code will be optimized as part of https://r3-cev.atlassian.net/browse/CORE-5271 anyway to use getReference and avoid loading data if possible. I haven't got a solution working for it yet so didn't include in this PR.

driessamyn commented on components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/DatabaseChunkPersistence.kt:
```diff
@@ -352,57 +369,38 @@ class DatabaseChunkPersistence(private val entityManagerFactory: EntityManagerFa
 
-    private fun createOrUpdateCpkMetadataEntities(cpi: Cpi, existingCpks: Map<CpkKey, CpiCpkEntity>): List<Pair<String, CpiCpkEntity>> {
-        return cpi.cpks.map { cpk ->
+    private fun createOrUpdateCpiCpkEntities(cpi: Cpi, existingCpks: Map<CpkKey, CpiCpkEntity>): Set<CpiCpkEntity> {
```
Minor thing: I think it's worth maybe breaking this class up a bit. If feels like the CPI lifecycle specific things could be on their own?

sprintengine replied:
Yeah I have a separate PR for later to do that - https://github.com/corda/corda-runtime-os/pull/1495

driessamyn commented on components/virtual-node/sandbox-group-context-service/src/test/kotlin/net/corda/sandboxgroupcontext/impl/Helpers.kt:
```diff
@@ -18,16 +18,23 @@ import org.mockito.kotlin.doReturn
 object Helpers {
-    private fun mockCpkMetadata(mainBundle: String, name: String, version: String): CpkMetadata
-    = mockCpkMetadata(mainBundle, emptyList(), name, version)
+    private fun mockCpkMetadata(
```
Bit unrelated to this PR and I don't suggest changing it here, but I wonder if we should think about putting these test object generators somewhere central in the `testing` folder as I think we do this type of thing in a few places.
Maybe alongside `testing/cpi-info-read-service-fake` ?
I'd be interested in other opinions on this.

barrylapthorn replied:
We should move more generators into the `testing` folder, though in this case I think in the new "packaging core"  these objects are easily constructable so do not need to be mocked (IIRC v1 cpi/cpk instances could only be easily generated by parsing an actual CPI).

driessamyn commented on libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiMetadataEntity.kt:
```diff
@@ -103,21 +103,7 @@ data class CpiMetadataEntity(
-                        cpkFileChecksum = cpiCpkEntity.metadata.cpkFileChecksum
-                    )
-                }.toSet()
+                cpks
```
What was the reason for pulling this out of this class?

I thought `CpiCpk` was just a mapping entity, so could stay internal to this module.

sprintengine replied:
I pass in `Set<CpiCpkEntity>` instead of `List<Pair<String, CpkMetadata>>` because we need to set the correct version on the `CpiCpkEntity`. Alternatively we could have passed in a `List<Triple<fileName, cpkMetadata, cpiCpkVersion>>` but I thought that was a bit messy.

EntityVersion on cpiCpk gets incremented because it has the `cpkFileChecksum` which gets updated when we update the cpkMetadata. Though I'm not sure its actually necessary to have this column on cpi_cpk.

driessamyn replied:
Maybe that's a leftover from when we were joining using the checksum. It's probably enough for the checksum to be on CpkMetadata only.

sprintengine replied:
Added to https://r3-cev.atlassian.net/browse/CORE-5271 to take care of

barrylapthorn commented on applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/virtualnode/VirtualNodeRpcTest.kt:
```diff
@@ -222,76 +223,134 @@ class VirtualNodeRpcTest {
-            // Depends on the flows in the cpi
-            val className = "net.cordapp.testing.calculator.CalculatorFlow"
-            val clientRequestId = Random.nextInt()
+            // BUG:  returning "OK" feels 'weakly' typed
```
Is this my comment cut'n'pasted?  I commented in this smoke test quite a bit about "odd" things I encountered as potential "customer" that would do something similiar (write an app against our REST api).

sprintengine replied:
I think it's just appearing as a change because I moved some code around when creating function to runCalculatorFlow()

driessamyn replied:
Worth noting that I believe @owenstanford is refactoring these to take flow invocation out of the vnode smoketests, into their own set.

sprintengine replied:
I think that change has already been merged, I'm rebasing with it now. The VirtualNodeRpcTest has a few conflicts such as no longer using calculator CPK. 

I think best course of action is to split out a few cpi upload specific tests into separate smoketest class (with order 30), and there I can upload the calculators and test them out. I will still need to call flows as part of this test, was the intention to have FlowTestUtils the only test that invokes flows? I guess as long as the flow tests run first that's OK (they are defined with `@Order(20)` at the class level.

barrylapthorn commented on applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/virtualnode/VirtualNodeRpcTest.kt:
```diff
@@ -222,76 +223,134 @@ class VirtualNodeRpcTest {
-                condition { it.code == 200 && it.toJson()["flowStatus"].textValue() == "COMPLETED" }
-            }.toJson()
+            // Check that timestamp for CPK been updated
+            // Cannot use `assertWithRetry` as there is a strict type `Instant`
```
You could extend the "DSL" if you wanted with a new `assertAfterDelay` - but this is probably just as readable.

The `assertWithRetry` only came about because clearly anything we do with REST takes some time so we need to back off and give the corda cluster time to do "stuff".

sprintengine replied:
Yeah that would make the test a bit nicer. I can do that as a follow up just to get this PR merged as it already quite a large footprint.

barrylapthorn replied:
Yeah, just a comment, not "changes requested" thing.

conal-smith committed:
CORE-4868 - Force CPI upload - Implement changes to ensure flow sandbox cache eviction

This change touches several APIs relevant for ensuring that when a CPI is force uploaded, the flow sandbox cache will be updated with a CPI with the new CPK.

This involved changing the CpkReadService API, the VirtualNodeContext api and making several updates to CPx related entities to ensure optimistic locking versions are correct when we effectively overwrite the old CPK data.

A test in VirtualNodeRpcTest ensures a force uploaded CPI will allow the user to invoke flows from the new CPI and asserts the response from the new flow is as expected.

conal-smith committed:
CORE-4868 - Extract force upload smoke tests from VirtualNodeRpcTest.kt into its own smoke test class since this test class now uses a different CPI and holding identity. We can use a clean virtual node with a clean CPI using the calculators.

conal-smith committed:
CORE-4868 - Fix the method to get the CPI file checksum which was incorrect and causing smoke tests to fail. Fix ordering of tests to create CPI before the virtual node.

driessamyn commented on applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/Utils.kt:
```diff
@@ -1,10 +1,23 @@
 const val X500_ALICE = "CN=Alice, OU=Application, O=R3, L=London, C=GB"
-const val CPI_NAME = "flow-worker-dev"
+const val X500_CAROL = "CN=Carol, OU=Application, O=R3, L=London, C=GB"
+const val FLOW_WORKER_DEV_CPI_NAME = "flow-worker-dev"
+const val CALCULATOR_CPI_NAME = "calculator"
```
Do we need 2 smoketest CPIs?
These are just smoketests, so should be a single happy path testcase through the system. To reduce maintenance overhead, I think we should reduce the number of test apps etc.

barrylapthorn replied:
I think Owen added the `flow-worker-dev` in a previous PR - not sure about this one though.

sprintengine replied:
Discussed with Viktor and Dries - I'll add a flow to flow-worker-dev that I can alter in a force-uploaded CPI and make an assertion on the return string. I'll revert a few changes here to the smoke tests.

driessamyn commented on applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/cpi/CpiForceUploadRpcTest.kt:
```diff
@@ -0,0 +1,202 @@
+            endpoint(CLUSTER_URI, USERNAME, PASSWORD)
+
+            val requestId = cpiUpload(ADDITION_CALCULATOR_CPB, GROUP_ID).let { it.toJson()["id"].textValue() }
+            assertThat(requestId).withFailMessage(ERROR_IS_CLUSTER_RUNNING).isNotEmpty
+
+            val json = assertWithRetry {
+                command { cpiStatus(requestId) }
+                condition { it.code == 200 && it.toJson()["status"].textValue() == "OK" }
+            }.toJson()
+
+            val cpiHash = json["checksum"].textValue()
+            assertThat(cpiHash).isNotNull.isNotEmpty
+
+            assertThat(cpiHash!!.length)
+                .withFailMessage("Short code length of wrong size - likely this test needs fixing")
+                .isEqualTo(12)
+
+            val cpiChecksum = getCpiChecksum(CALCULATOR_CPI_NAME)
+
+            assertThat(cpiChecksum).isNotNull.isNotEmpty
+            assertThat(cpiHash).withFailMessage(ERROR_CPI_NOT_UPLOADED).isNotNull
+            assertThat(cpiChecksum).isEqualTo(cpiHash)
+        }
+    }
+
+    @Test
+    @Order(31)
+    fun `create a clean virtual node`() {
+        cluster {
+            endpoint(CLUSTER_URI, USERNAME, PASSWORD)
+            val hash = getCpiChecksum(CALCULATOR_CPI_NAME)
+
+            val vNodeJson = assertWithRetry {
+                command { vNodeCreate(hash, X500_CAROL) }
+                condition { it.code == 200 }
+                failMessage(ERROR_HOLDING_ID)
+            }.toJson()
+
+            assertThat(vNodeJson["holdingIdHash"].textValue()).isNotNull.isNotEmpty
+        }
```
I would prefer it if we didn't have overlapping test cases.
I think these could be additions to the existing tests.
If it's a matter of separating out tests in their own classes, we can [order the test class execution](https://junit.org/junit5/docs/snapshot/user-guide/index.html#writing-tests-test-execution-order-classes) similar to how we already order the method execution.

sprintengine replied:
We now have a few extra tests in VirtualNodeRpcTest to test the force upload functionality. It was also required we run flows from this test, hence why it uses a few functions from the FlowTestUtils utility class.

conal-smith committed:
CORE-4868 - Add a flow to the Smoke test CPI instead of inserting a new CPI for calculator, make a new CPI artifact containing a changed CPK and make assertions about this ReturnAStringFlow.kt.

driessamyn reviewed (approved):
LGTM
